File: src/main/java/com/nuecho/genesys/cli/compare/CompareAgents.java
package com.nuecho.genesys.cli.compare;

import static com.google.common.collect.Lists.newArrayList;
import static com.nuecho.genesys.cli.GenesysServices.releaseConfigurationService;
import static java.util.stream.Collectors.toList;

import com.genesyslab.platform.applicationblocks.com.IConfService;
import com.genesyslab.platform.applicationblocks.com.objects.CfgPerson;
import com.genesyslab.platform.applicationblocks.com.queries.CfgPersonQuery;
import com.genesyslab.platform.configuration.protocol.types.CfgFlag;
import difflib.DiffUtils;
import difflib.Patch;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.Callable;
import picocli.CommandLine;

@CommandLine.Command(
  name = "agents",
  description = "Compare Genesys agent configuration between two instances."
)
public final class CompareAgents implements Callable<Void> {
  @CommandLine.ParentCommand private Compare compare;

  @Override
  public Void call() throws Exception {
    IConfService service = null;
    try {
      service = compare.connect();

      CfgPersonQuery query = new CfgPersonQuery();
      query.setIsAgent(CfgFlag.CFGTrue.ordinal());
      Collection<CfgPerson> persons = service.retrieveMultipleObjects(CfgPerson.class, query);
      List<String> left = persons.stream().map(person -> person.getUserName()).collect(toList());
      List<String> right = newArrayList(left);

      /* Sample test, to use with 192.168.129.85 server
      right.remove(10);
      right.remove(10);
      right.remove(10);

      right.remove(50);
      right.remove(50);
      right.remove(50);

      right.set(5, "testerpatate");
      */

      Patch diff = DiffUtils.diff(left, right);
      System.out.println(diff.getDeltas());
    } finally {
      releaseConfigurationService(service);
    }

    return null;
  }
}

File: src/main/kotlin/com/nuecho/genesys/cli/CompareAgents.kt
package com.nuecho.genesys.cli

import com.genesyslab.platform.applicationblocks.com.IConfService
import com.genesyslab.platform.applicationblocks.com.objects.CfgPerson
import com.genesyslab.platform.applicationblocks.com.queries.CfgPersonQuery
import com.genesyslab.platform.configuration.protocol.types.CfgFlag
import difflib.DiffUtils
import picocli.CommandLine
import java.util.concurrent.Callable

@CommandLine.Command(name = "agents", description = ["Compare Genesys agent configuration between two instances."])
class CompareAgents : BasicCommand(), Callable<Void> {
    @CommandLine.ParentCommand
    private val compare: Compare? = null

    @Throws(Exception::class)
    override fun call(): Void? {
        var service: IConfService? = null
        try {
            service = compare!!.connect()

            val query = CfgPersonQuery()
            query.isAgent = CfgFlag.CFGTrue.ordinal()
            val persons = service.retrieveMultipleObjects(CfgPerson::class.java, query)
            val left = persons.map { person -> person.userName }
            val right = left.toMutableList()

            // Sample test, to use with 192.168.129.85 server
            right.removeAt(10)
            right.removeAt(10)
            right.removeAt(10)

            right.removeAt(50)
            right.removeAt(50)
            right.removeAt(50)

            right[5] = "testerpatate"

            val diff = DiffUtils.diff(left, right)
            println(diff.deltas)
        } finally {
            GenesysServices.releaseConfigurationService(service)
        }

        return null
    }
}
